php-frontend/src/main/java/org/sonar/php/utils/LiteralUtils.java

```diff
@@ -33,4 +36,146 @@ public static long longLiteralValue(String literalValue) {
     return Long.decode(value);
   }
 
+  // https://www.php.net/manual/en/language.types.string.php
+  public static String stringLiteralValue(String literalValue) {
+    return StringLiteralParser.stringValue(literalValue);
+  }
+
+  private abstract static class StringLiteralParser {
+
+    private final String valueWithoutDelimiters;
+
+    static String stringValue(String literalValue) {
+      String valueWithoutDelimiters = literalValue.substring(1, literalValue.length() - 1);
+      if (valueWithoutDelimiters.indexOf('\\') == -1) {
+        return valueWithoutDelimiters;
+      }
+      StringLiteralParser parser = (literalValue.charAt(0) == '\'')
+        ? new SingleQuotedStringLiteralParser(valueWithoutDelimiters)
+        : new DoubleQuotedStringLiteralParser(valueWithoutDelimiters);
+      return parser.stringValue();
+    }
+
+    StringLiteralParser(String valueWithoutDelimiters) {
+      this.valueWithoutDelimiters = valueWithoutDelimiters;
+    }
+
+    String stringValue() {
+      StringBuilder stringValue = new StringBuilder();
+      boolean isInEscapeSequence = false;
+      int i = 0;
+      while (i < valueWithoutDelimiters.length()) {
+        char c = valueWithoutDelimiters.charAt(i);
+        if (isInEscapeSequence) {
+          String remainder = valueWithoutDelimiters.substring(i);
+          int escapeSequenceLength = handleEscapeSequence(stringValue, c, remainder);
+          isInEscapeSequence = false;
+          i += escapeSequenceLength - 1;
+        } else {
+          if (c == '\\') {
+            isInEscapeSequence = true;
+          } else {
+            stringValue.append(c);
+          }
+          i++;
+        }
+      }
+      return stringValue.toString();
+    }
+
+    /**
+     * @return the total number of characters in the handled escape sequence
+     */
+    abstract int handleEscapeSequence(StringBuilder stringValue, char charAfterBackslash, String remainder);
+  }
+
+  private static class SingleQuotedStringLiteralParser extends StringLiteralParser {
+
+    SingleQuotedStringLiteralParser(String valueWithoutDelimiters) {
+      super(valueWithoutDelimiters);
+    }
+
+    @Override
+    int handleEscapeSequence(StringBuilder stringValue, char charAfterBackslash, String remainder) {
+      if (charAfterBackslash == '\'') {
+        stringValue.append('\'');
+      } else if (charAfterBackslash == '\\') {
+        stringValue.append('\\');
+      } else {
+        stringValue.append('\\').append(charAfterBackslash);
+      }
+      return 2;
+    }
+  }
+
+  private static class DoubleQuotedStringLiteralParser extends StringLiteralParser {
+
+    DoubleQuotedStringLiteralParser(String valueWithoutDelimiters) {
+      super(valueWithoutDelimiters);
+    }
+
+    @Override
+    int handleEscapeSequence(StringBuilder stringValue, char charAfterBackslash, String remainder) {
+      switch (charAfterBackslash) {
+        case '\\':
+          stringValue.append("\\");
```

Comment:
```suggestion
          stringValue.append('\\');
```
Maybe also a char instead of a string as above.